source/scpd/Cpp.d

```diff
@@ -373,10 +373,28 @@ extern(C++, (StdNamespace)) extern(C++, class) struct vector (T, Alloc = allocat
             foreach (idx; 0 .. len)
             {
                 auto entry = deserializeFull!(QT.ElementType)(data, opts);
-                push_back(cast() ret, entry);
+                push_back_std_vector(cast() ret, entry);
```

Comment:
This should work without introducing `push_back_std_vector`:

```suggestion
                auto val = cast()ret;
                val.push_back(entry);
```

Reply:
Actually this cast should be moved below the `Unqual!(vector!(Unqual!(QT.ElementType))) ret;` call, and then replace `return () @trusted { return cast(QT) ret; }();` so it returns that value.

Reply:
The suggested code fails in the unit test.

```Module tests failed: agora.test.GenesisBlock
core.exception.AssertError@source/agora/test/GenesisBlock.d(99): Check condition failed after timeout of 2 secs and 20 attempts
----------------
??:? void agora.utils.Test.retryFor(lazy bool, core.time.Duration, lazy immutable(char)[], immutable(char)[], ulong) [0x109617b7e]
??:? void agora.test.GenesisBlock.__unittest_L75_C1() [0x109648973]
??:? void agora.test.Base.customModuleUnitTester().runTest(agora.test.Base.customModuleUnitTester().ModTest) [0x10964b901]
??:? int agora.test.Base.customModuleUnitTester().__foreachbody4(ref agora.test.Base.customModuleUnitTester().ModTest) [0x10964bc40]
??:? void std.parallelism.ParallelForeach!(agora.test.Base.customModuleUnitTester().ModTest[]).ParallelForeach.opApply(scope int delegate(ref agora.test.Base.customModuleUnitTester().ModTest)).doIt() [0x1099cb121]
??:? void std.parallelism.TaskPool.doJob(std.parallelism.AbstractTask*) [0x109f0c830]
??:? void std.parallelism.TaskPool.startWorkLoop() [0x109f0c9da]
??:? thread_entryPoint [0x109f92a24]
??:? _pthread_start [0x7fff6624fe64]```

Reply:
Can you show me the diff that you made?

Reply:
```
diff --git a/source/scpd/Cpp.d b/source/scpd/Cpp.d
index 26f650d2..e61f6095 100644
--- a/source/scpd/Cpp.d
+++ b/source/scpd/Cpp.d
@@ -373,7 +373,8 @@ extern(C++, (StdNamespace)) extern(C++, class) struct vector (T, Alloc = allocat
             foreach (idx; 0 .. len)
             {
                 auto entry = deserializeFull!(QT.ElementType)(data, opts);
-                push_back_std_vector(cast() ret, entry);
+                auto val = cast()ret;
+                val.push_back(entry);
             }
             return () @trusted { return cast(QT) ret; }();
         }
diff --git a/source/scpd/types/Utils.d b/source/scpd/types/Utils.d
index 183cca42..ee3d9b9f 100644
--- a/source/scpd/types/Utils.d
+++ b/source/scpd/types/Utils.d
@@ -60,4 +60,3 @@ public void push_back(T, VectorT) (ref VectorT this_, ref T value) @safe pure no
 // Workaround for Dlang issue #20805
 public void push_back_vec (void*, const(void)*) @safe pure nothrow @nogc;
 public VectorT duplicate(VectorT)(ref const VectorT this_) @safe pure nothrow @nogc;
-public void push_back_std_vector(T, VectorT) (ref VectorT this_, ref T value) @safe pure nothrow @nogc;
diff --git a/source/scpp/extra/DUtils.cpp b/source/scpp/extra/DUtils.cpp
index 015fcee3..be80676e 100644
--- a/source/scpp/extra/DUtils.cpp
+++ b/source/scpp/extra/DUtils.cpp
@@ -31,7 +31,6 @@ opaque_vec<> XDRToOpaque(const stellar::SCPStatement& param)
 #define PUSHBACKINST1(T) template void push_back<T, xvector<T>>(xvector<T>&, T&);
 #define PUSHBACKINST2(T, VT) template void push_back<T, VT>(VT&, T&);
 #define PUSHBACKINST3(T, V) template void push_back<T, V<T>>(V<T>&, T&);
-#define PUSHBACKINST_STD_VECTOR(T, VT) template void push_back_std_vector<T, VT>(VT&, T&);

 PUSHBACKINST2(const PublicKey, xvector<PublicKey>)
 PUSHBACKINST3(xvector<unsigned char>, std::vector)
@@ -57,12 +56,6 @@ PUSHBACKINST3(PublicKey, std::vector)
 PUSHBACKINST3(SCPEnvelope, std::vector)
 PUSHBACKINST3(SCPQuorumSet, std::vector)

-PUSHBACKINST_STD_VECTOR(PublicKey   , std::vector<PublicKey>)
-PUSHBACKINST_STD_VECTOR(SCPQuorumSet, std::vector<SCPQuorumSet>)
-PUSHBACKINST_STD_VECTOR(SCPEnvelope , std::vector<SCPEnvelope>)
-PUSHBACKINST_STD_VECTOR(xvector<unsigned char>, std::vector<xvector<unsigned char>>)
-PUSHBACKINST_STD_VECTOR(unsigned char, std::vector<unsigned char>)
-
 template opaque_vec<> duplicate<opaque_vec<>>(opaque_vec<> const&);

 #define CPPSETFOREACHINST(T) template int cpp_set_foreach<T>(void*, void*, void*);
diff --git a/source/scpp/extra/DUtils.h b/source/scpp/extra/DUtils.h
index 1cb43263..2b156389 100644
--- a/source/scpp/extra/DUtils.h
+++ b/source/scpp/extra/DUtils.h
@@ -67,12 +67,6 @@ void push_back(VectorT& this_, T& value)
     this_.push_back(value);
 }

-template<typename T, typename VectorT>
-void push_back_std_vector(VectorT& this_, T& value)
-{
-    this_.push_back(value);
-}
-
```

Reply:
Try this:

```diff
diff --git a/source/scpd/Cpp.d b/source/scpd/Cpp.d
index 26f650d2..81dd47ea 100644
--- a/source/scpd/Cpp.d
+++ b/source/scpd/Cpp.d
@@ -373,7 +373,7 @@ extern(C++, (StdNamespace)) extern(C++, class) struct vector (T, Alloc = allocat
             foreach (idx; 0 .. len)
             {
                 auto entry = deserializeFull!(QT.ElementType)(data, opts);
-                push_back_std_vector(cast() ret, entry);
+                ret.push_back(entry);
             }
             return () @trusted { return cast(QT) ret; }();
         }
```

It works on MacOS, and I hope it will work on Windows too.

Reply:
I'm not sure why that `cast` was there in the first place, I think it was likely just leftover by mistake. Because the `Unqual` already makes the vector mutable.

Reply:
I try it! me too!

Reply:
I trusted that code at first.
However, the Vector was wrapped twice in the override push_back.